Proj 270 get feeder resolve all references

kgreav opened the pull request:
@charlta @gcarwardine 

So this PR does a few things but notably makes get_feeder resolve all references rather than the previous subset of types.
From what I understand Carlos isn't currently blocked on this as he has a workaround at the moment, so it's not a huge rush. It requires some thought and may be slightly controversial (although hopefully not)

The other big change is around unresolved references - I've changed them to be indexed in two ways: "from" and "to". 
The "to" map is an index over to_mrid (as it was previously) and there is now a new "from" map which has the reverse. Both maps always contain the same data, they are just indexed differently.

This allows for some more efficient querying of unresolved references, as you can easily get all the references for an object by its mRID, rather than having to iterate over all the unresolved references.

The drawback is the additional memory cost. From my experimenting this is only really a significant cost when you pull in a Feeder, and it's not that bad. We essentially double the amount of references held in memory, but this is only for a limited time while you are resolving references, and the memory use of the maps drops twice as fast as it used to as you resolve references.

For my test feeder of ~18000 objects the memory usage was insignificant, so I'd only expect it to be problematic if we started getting up to the 100 thousands or millions on a single feeder.

I also made the maps values sets rather than lists to enhance lookup times - which previously was quite expensive when deferring references or adding new objects.
And finally I made `UnresolvedReference.from` an mRID rather than an `IdentifiedObject` - this was just during my debugging and it's not strictly needed, but I've left it in there for the moment as I think it's cleaner/easier to understand. The drawback is that you can't immediately resolve a reference for an object that's _not_ in the service, however I don't think we should really be allowing that anyway as it could be misleading for users.

I know it's all in python but you should be able to get the gist of it pretty easily, the important parts are in base_service.py, network_consumer.py.

ryanjcoleman commented:
Task linked: [PROJ-270 Get feeder resolves all immedaite references in Python SDK](https://app.clickup.com/t/6929263/PROJ-270)

gcarwardine commented on changelog.md:
```diff
@@ -5,6 +5,7 @@
 
 ##### New Features
```
Probably worth also updating the changelog with the consumer changes.

gcarwardine replied:
Also noting the UUID is not valid for MRIDs anymore.

gcarwardine commented on docs/docs/consumer.mdx:
```diff
@@ -85,7 +85,15 @@ async def get_with_base_voltage(service: NetworkService, client: NetworkConsumer
-    mrids = service.get_unresolved_reference_mrids(resolver.ce_base_voltage(equipment))
-    if mrids:
-        await client.get_identified_object(service, mrids[0])
+    # Get all unresolved references
```
Should update this example to be more sensible. People will end up pulling in too much stuff.

gcarwardine commented on src/zepben/evolve/services/common/base_service.py:
```diff
@@ -208,12 +243,33 @@ def get_unresolved_reference_mrids(self, bound_resolvers: Union[BoundReferenceRe
+                    seen.add(ref.to_mrid)
+                    yield ref.to_mrid
+
+    def get_unresolved_reference_mrids_from(self, mrid: str) -> Generator[str, None, None]:
```
So based on our discussion I think get_unresolved_reference_mrids_from and get_unresolved_reference_mrids_to should not return MRIDs but should return the UnresovledReferences themselves. That way you can see what type the unresolved id is. It's probably fine to leave the get_unresolved_reference_mrids_by_resolver as is because you know the type due to the resolver, but I wonder if return the UnresolvedReference instsance is better to just be consistent?

gcarwardine commented on src/zepben/evolve/streaming/get/network_consumer.py:
```diff
@@ -130,98 +130,109 @@ async def y():
+        # however we don't know how long this tree is until we've resolved everything. So we start by resolving all the equipment, then we iteratively
+        # descend the tree by checking for unresolved references for the objects we just resolved in the previous iteration.
+        to_resolve = self._get_unresolved_mrids(service, res.value.keys())
+        while True:
```
So this needs to be reworked as discussed. We should resolve references unless they represent some form of connectivity that will blow out to resolving everything. Initial items to restrict:
* Feeders should not automatically resolve equipment
* ConnectivityNodes should not automatically resolve its terminals.

charlta replied:
@gcarwardine thinking more on this, another way that this can creep is via `UsagePoint` and `Meter` or via `OperationalRestriction`. There may be more but resolving all relationships may not be the best way forward, although the creep in these casses will be much lower as the connectivity will not be traversed

kgreav replied:
yeah i had a think about this and came to the conclusion that if we exclude the ConnectivityNode -> Terminal relationship when resolving we shouldn't creep beyond what's actually included in the OperationalRestriction/UsagePoint/etc, which you'd probably want anyway if you were pulling in the feeder?

charlta commented on src/zepben/evolve/streaming/get/network_consumer.py:
```diff
@@ -130,103 +152,139 @@ async def y():
+        # also do the same for the resolutions, and so on so forth until the entire tree of references originating from the Feeder has been resolved,
+        # however we don't know how long this tree is until we've resolved everything. So we start by resolving all the equipment, then we iteratively
+        # descend the tree by checking for unresolved references for the objects we just resolved in the previous iteration.
+        to_resolve = self._get_unresolved_mrids(service, res.result.value.keys(), _EXCLUDED_GET_FEEDER)
```
Why are other feeders being excluded? This is information that should be brought back and with the removal of equipment ID's is not going to cause the creep it used to

kgreav replied:
do we want the hierarchy to come back with get_feeder though? i was thinking the typical usage pattern would be to get the network hierarchy and then from that get a feeder

gcarwardine replied:
@kgreav I'm a bit confused why we would pull the whole hierarchy back to get the feeder when the specific feeder has been requested?

kgreav replied:
no, i was saying that would be a typical usage pattern for a user and that we wouldn't want get_feeder to return/fetch the network hierarchy.

anyway this is all moot now, feeders wont be excluded, but you'll only get other feeders if some equipment belongs to them (but you won't get those feeders equipment).

charlta replied:
@gcarwardine The entire hierachy will be pulled in if there are any open points on the requested feeder that have a different feeder on the other side of the switch due to those other feeders being added to the recursive resolve loop

@kgreav Given that it is highly likely that the feeder has open points with at least one other feeder you can porbably just include the selected feeder in the recursive loop rather than having a second lookup for the substation only

kgreav replied:
yep, good point, i've moved it into the loop.

gcarwardine commented:
@kgreav in your initial comment you wrote
> And finally I made UnresolvedReference.from an mRID rather than an IdentifiedObject - this was just during my debugging and it's not strictly needed, but I've left it in there for the moment as I think it's cleaner/easier to understand. The drawback is that you can't immediately resolve a reference for an object that's not in the service, however I don't think we should really be allowing that anyway as it could be misleading for users.

That doesn't sound right to me. You've left `BoundReferenceResolver` with the identified object. So things can resolve immediately in resolveOrDeferReference. What it would do is in the case of a deferred resolution, if you add the "to" before the "from" to the service, the add of the "to" object would not be able to resolve it's reference to the from. You would then end up with unresolved references in the service that would never resolve. This is unlikely to happen but an undesirable outcome. Do you have any ideas how to address this?

kgreav commented:
Hm, not sure if I completely follow, but yes this introduces a considerable problem with doing things out of order. e.g:
(where x and y have a reference to each other)
resolve_or_defer_reference(x)
resolve_or_defer_reference(y)
add(x) - fails

Anyway, i've reverted that change and put the references back (and also added a test to catch this).

kgreav commented:
Alright, this is ready for review again along with evolve-grpc. There is one concern - with this implementation of get_feeder we pull in all substations and feeders, as SubgeographicalRegion still has the substation mRIDs and Substation still has Feeder mRIDs. thus get_feeder will effectively give you the whole hierarchy. I don't think this is a problem really, but if either of you are concerned we'd have to fix it by doing the same thing for subgeo's and substations.

kgreav commented:
when implementing the JVM side of this i came across a few things i missed in here (including a test), nothing major but i'll fix it up on thursday assuming the JVM SDK doesn't require any further changes.

gcarwardine reviewed (approved):
Once you are happy this marries up with the JVM changes this looks good to me. We need to make sure we get all this functionality tested in a real case sooner rather than later though so we can squash any bugs while this is all still fresh.
